### PR TITLE
Version 4.1.0

### PR DESCRIPTION
**Pull request by mindstorm38**
### Changes
- *API*: LWJGL fix can now support all present and future 3.3.* versions, including 3.2.3 like before (#181)
- *API*: Added **unstable** support for NeoForge repository in `ForgeVersion` (#180, thanks @MisileLab)
- *API*: Added a default User-Agent for HTTP requests *(not for downloads)* of `portablemc/<version>`
- *CLI*: LWJGL fix is no longer restricted to specific values, you can enter any value and the launcher will raise an error (download error if the version is not yet available) if it is not supported (#181)
- *CLI*: Added support for NeoForge launch with `neoforge:` id and associated `--neoforge-prefix` argument (#180)
- *CLI*: Notice about the `--login` argument when using thje `login` command (#182, thanks @superniels09)
- *Doc*: Improved README

### CODE REVIEW

**Comment by MisileLab** [comment]
I'm using portablemc as primary launcher so neoforge support would be great

**Comment by mindstorm38** [comment]
> I'm using portablemc as primary launcher so neoforge support would be great

The CLI support will be stable in the next update!

File: portablemc/forge.py [hunk 1]
```diff
@@ -18,52 +20,74 @@
 from typing import Dict, Optional, List
 
 
+_FORGE_REPO = "https://maven.minecraftforge.net/net/minecraftforge/forge"
+_NEO_FORGE_REPO = "https://maven.neoforged.net/releases/net/neoforged/forge"
```
Comment (mindstorm38) [inline portablemc/forge.py:24]: Verified unstable API

File: portablemc/forge.py [hunk 2]
```diff
@@ -18,52 +20,74 @@
...
+
+
 class ForgeVersion(Version):
     
     def __init__(self, forge_version: str = "release", *,
         context: Optional[Context] = None,
         prefix: str = "forge",
+        _forge_repo: str = _FORGE_REPO,
```
Comment (mindstorm38) [inline portablemc/forge.py:32]: Defaults to forge repo, the argument is unstable

File: portablemc/forge.py [hunk 3]
```diff
@@ -102,12 +126,12 @@ def _fetch_version(self, version: VersionHandle, watcher: Watcher) -> None:
         install_jar = None
         for suffix in suffixes:
             try:
-                install_jar = request_install_jar(f"{self.forge_version}{suffix}")
+                install_jar = request_install_jar(f"{self.forge_version}{suffix}", _repo=self._forge_repo)
```
Comment (mindstorm38) [inline portablemc/forge.py:129]: Internally using an unstable API is verified

File: portablemc/forge.py [hunk 4]
```diff
@@ -397,10 +421,11 @@ class ForgeResolveEvent:
     The 'alias' attribute specifies if an alias version is being resolved, if false the
     resolving has finished and we'll try to install the given version.
     """
-    __slots__ = "forge_version", "alias"
-    def __init__(self, forge_version: str, alias: bool) -> None:
+    __slots__ = "forge_version", "alias", "_forge_repo"
+    def __init__(self, forge_version: str, alias: bool, *, _forge_repo: str) -> None:
         self.forge_version = forge_version
         self.alias = alias
+        self._forge_repo = _forge_repo
```
Comment (mindstorm38) [inline portablemc/forge.py:428]: Verified unstable API for the event

File: portablemc/forge.py [hunk 5]
```diff
@@ -416,16 +441,38 @@ class ForgePostProcessedEvent:
...
 
-def request_maven_versions() -> List[str]:
+def _request_neoforge_version(game_version: str) -> Optional[str]:
+    """Request a neoforge version from a game version. Because of the NeoForge API 
+    returning any matching version number, this function will also check that the returned
+    version is of the right game version.
+    """
+    try:
+        # NOTE: For now we don't sanitize the parameter.
+        url = f"https://maven.neoforged.net/api/maven/latest/version/releases/net%2Fneoforged%2Fforge?filter={game_version}"
```
Comment (mindstorm38) [inline portablemc/forge.py:458]: TODO: Sanitize this parameter

File: portablemc/forge.py [hunk 6]
```diff
@@ -416,16 +441,38 @@ class ForgePostProcessedEvent:
...
+        return None
+
+
+def request_maven_versions(*, _repo: str = _FORGE_REPO) -> List[str]:
     """Internal function that parses maven metadata of forge in order to get all 
     supported forge versions.
     """
 
-    text = http_request("GET", "https://maven.minecraftforge.net/net/minecraftforge/forge/maven-metadata.xml", 
+    text = http_request("GET", f"{_repo}/maven-metadata.xml",
```
Comment (mindstorm38) [inline portablemc/forge.py:475]: Verified unstable API, existing code won't break because it defaults to the same repo

File: portablemc/forge.py [hunk 7]
```diff
@@ -447,11 +494,10 @@ def request_maven_versions() -> List[str]:
     return versions
 
 
-def request_install_jar(version: str) -> ZipFile:
+def request_install_jar(version: str, *, _repo: str = _FORGE_REPO) -> ZipFile:
     """Internal function to request the installation JAR file.
     """
-    # print(f"https://maven.minecraftforge.net/net/minecraftforge/forge/{version}/forge-{version}-installer.jar")
-    res = http_request("GET", f"https://maven.minecraftforge.net/net/minecraftforge/forge/{version}/forge-{version}-installer.jar",
+    res = http_request("GET", f"{_repo}/{version}/forge-{version}-installer.jar",
```
Comment (mindstorm38) [inline portablemc/forge.py:500]: Verified unstable API, do not break existing code

File: portablemc/cli/__init__.py [hunk 8]
```diff
@@ -447,17 +456,22 @@ def cmd_start_handler(ns: StartNs, kind: str, parts: List[str]) -> Optional[Vers
...
         return constructor(version, parts[1] if len(parts) == 2 else None, context=ns.context, prefix=prefix)
     
-    elif kind == "forge":
+    elif kind in ("forge", "neoforge"):
         if len(parts) != 1:
             return None
-        return ForgeVersion(version, context=ns.context, prefix=ns.forge_prefix)
+        repo = _FORGE_REPO if kind == "forge" else _NEO_FORGE_REPO
+        prefix = ns.forge_prefix if kind == "forge" else ns.neoforge_prefix
+        return ForgeVersion(version, context=ns.context, prefix=prefix, _forge_repo=repo)
```
Comment (mindstorm38) [inline portablemc/cli/__init__.py:464]: Using unstable API is legal in CLI

File: portablemc/cli/__init__.py [hunk 9]
```diff
@@ -748,10 +762,11 @@ def fabric_resolve(e: FabricResolveEvent) -> None:
                 ns.out.finish()
         
         def forge_resolve(e: ForgeResolveEvent) -> None:
+            api = "forge" if e._forge_repo == _FORGE_REPO else "neoforge"
             if e.alias:
-                ns.out.task("..", "start.forge.resolving", version=e.forge_version)
+                ns.out.task("..", "start.forge.resolving", api=api, version=e.forge_version)
             else:
-                ns.out.task("OK", "start.forge.resolved", version=e.forge_version)
+                ns.out.task("OK", "start.forge.resolved", api=api, version=e.forge_version)
```
Comment (mindstorm38) [inline portablemc/cli/__init__.py:769]: Using unstable API is legal in CLI

**Comment by MisileLab** [comment]
@mindstorm38 I think neoforge:1.20.1-latest (or other version) not working

**Comment by mindstorm38** [comment]
Remove the -latest, it's not needed for neoforge since there no latest/recommended distinction.

**Pull request merged**